File: Data Structures & Algorithms/target-sum/submission-7.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        memo = {}
        def dfs(i: int, current: int) -> int:
            if i == len(nums) and current == 0:
                return 1
            if i == len(nums):
                return 0
            if (i, current) in memo:
                return memo[(i, current)]

            memo[(i, current)] = dfs(i + 1, current - nums[i]) + dfs(i + 1, current + nums[i])
            return memo[(i, current)]

        return dfs(0, target)




#                         (0, 2)
#             (1,0)                   (1,4)
#     (2,-2)          (2,2)       (2,2)    (2,6)
# (3,0)  (3,-4)   (3,4)   (3,0) (3,4) (3,0)

File: Data Structures & Algorithms/combination-target-sum/submission-7.cpp
class Solution {
private:
    void dfs(int i, int sum, int target, vector<int>& nums, vector<int>& subset, vector<vector<int>>& result) {
        if (sum == target) {
            result.push_back(subset);
            return;
        }
        if (i == nums.size() || sum > target) return;

        subset.push_back(nums[i]);
        dfs(i, sum + nums[i], target, nums, subset, result);
        subset.pop_back();
        dfs(i + 1, sum, target, nums, subset, result);
    }

public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> result;
        vector<int> subset;
        dfs(0, 0, target, nums, subset, result);
        return result;
    }
};



                   